sycl/unittests/config/PreferredWGSizeConfigTests.cpp

```diff
@@ -0,0 +1,132 @@
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+// Tests that valid and invalid configuration values for
+// SYCL_REDUCTION_PREFERRED_WORKGROUP_SIZE behave as expected.
+
+#include <detail/config.hpp>
+#include <gtest/gtest.h>
+#include <regex>
+#include <sycl/sycl.hpp>
+
+// Sets the SYCL_REDUCTION_PREFERRED_WORKGROUP_SIZE configuration and forces
+// a reparse.
+void SetConfig(const char *Value) {
+#ifdef _WIN32
+  _putenv_s("SYCL_REDUCTION_PREFERRED_WORKGROUP_SIZE", Value);
+#else
+  setenv("SYCL_REDUCTION_PREFERRED_WORKGROUP_SIZE", Value, 1);
+#endif
+  sycl::detail::SYCLConfig<
+      sycl::detail::SYCL_REDUCTION_PREFERRED_WORKGROUP_SIZE>::reset();
+}
+
+// Gets the parsed value of the SYCL_REDUCTION_PREFERRED_WORKGROUP_SIZE
+// configuration for a given device-type.
+size_t GetConfigValue(sycl::info::device_type DevType) {
+  return sycl::detail::SYCLConfig<
+      sycl::detail::SYCL_REDUCTION_PREFERRED_WORKGROUP_SIZE>::get(DevType);
+}
+
+// Sets the SYCL_REDUCTION_PREFERRED_WORKGROUP_SIZE configuration and checks
+// the parsed values.
+void SetAndCheck(const char *ConfigValue, size_t CPUValue, size_t GPUValue,
+                 size_t AccValue) {
+  SetConfig(ConfigValue);
+  EXPECT_EQ(GetConfigValue(sycl::info::device_type::cpu), CPUValue)
+      << "Unexpected value for CPU with '" << ConfigValue << "'.";
+  EXPECT_EQ(GetConfigValue(sycl::info::device_type::gpu), GPUValue)
+      << "Unexpected value for GPU with '" << ConfigValue << "'.";
+  EXPECT_EQ(GetConfigValue(sycl::info::device_type::accelerator), AccValue)
+      << "Unexpected value for accelerator with '" << ConfigValue << "'.";
+}
+
+// Sets the SYCL_REDUCTION_PREFERRED_WORKGROUP_SIZE configuration and expects
+// a sycl::exception to be thrown with the specified error code.
+void SetAndExpectException(const char *ConfigValue,
+                           sycl::errc ExpectedErrorCode) {
+  try {
+    SetConfig(ConfigValue);
+    EXPECT_TRUE(false) << "Setting the config with '" << ConfigValue
+                       << "' unexpectedly succeeded.";
+  } catch (sycl::exception &E) {
+    EXPECT_EQ(E.code(), sycl::make_error_code(ExpectedErrorCode))
+        << "Exception thrown when setting the config with '" << ConfigValue
+        << "' does not have the expected error code.";
+  } catch (...) {
+    EXPECT_TRUE(false) << "Setting the config with '" << ConfigValue
+                       << "' throw a non-SYCL exception.";
+  }
+}
+
+// NOTE: All checks are kept in the same file to avoid potential multi-threading
+// from overwriting the program-wide configurations.
+TEST(ConfigTests, CheckPreferredWGSizeConfigProcessing) {
+  SetAndCheck("cpu:32", 32, 0, 0);
+  SetAndCheck("gpu:32", 0, 32, 0);
```

Comment:
So we can't set different values to different GPUs. I think this is fine for this PR, but maybe we'd need to change it later.

Any thoughts?

Reply:
We could extend it in the future indeed, but I would also like to see it as a property in the future instead.